### PR TITLE
fix(curriculum): Inefficient solution for the "Problem 78: Coin partitions" challenge.

### LINKED ISSUE #52772: Inefficient solution for the "Problem 78: Coin partitions" challenge.
### Describe the Issue

When trying to run tests on the local machine, one of them failed with the following error:
`AssertionError: expected 51999 to equal 55374`.
The test that fails is related to: `curriculum/challenges/english/18-project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions.md`

This behavior occurs on a freshly forked repository without any changes made.
But, strangely, in the web interface itself https://www.freecodecamp.org/learn/project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions when you click the `Run the Tests (Ctrl + Enter)` button or in the pipelines, this error does not exist.

### Affected Page

https://www.freecodecamp.org/learn/project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions

### Your code

Running the following command:
```
pnpm run test:curriculum
```

Causes the error:
```
oleksandr.tkachenko@SOF-X-L-2301 freeCodeCamp % pnpm run test:curriculum

> @freecodecamp/freecodecamp@0.0.1 test:curriculum /Users/oleksandr.tkachenko/WOLFRIEND/Dev/freeCodeCamp
> cd ./curriculum && pnpm test


> @freecodecamp/curriculum@0.0.0-next.4 test /Users/oleksandr.tkachenko/WOLFRIEND/Dev/freeCodeCamp/curriculum
> ts-node ./node_modules/mocha/bin/mocha.js --delay --exit --reporter progress --bail

⠧ Populate tests.


  23923 passing (4m)
  116 pending
  1 failing

  1) Check challenges
       Check challenges (english)
         project-euler-problems-1-to-100
           Problem 78: Coin partitions
             Check tests against solutions
               Solution 1 must pass the tests:
     Test text: <p><code>coinPartitions(1000000)</code> should return <code>55374</code>.</p>
  AssertionError: expected 51999 to equal 55374
      at eval (eval at ctx.onmessage (http://127.0.0.1:8080/js/test-evaluator-bb607458058e945d60ce.js:18292:26), <anonymous>:46:8)
      at ctx.onmessage (http://127.0.0.1:8080/js/test-evaluator-bb607458058e945d60ce.js:18292:26)



 ELIFECYCLE  Test failed. See above for more details.
 ELIFECYCLE  Command failed with exit code 1.

```

### Expected behavior

The test should be passed successfully without any errors.

### Screenshots

image

### System

- Device: MacBook Pro, 16-inch, 2021, Apple M1 Max.
- OS: macOS Sonoma 14.2.1 (23C71).
- Browser: N/A.
- Version: N/A/
- Node version: v20.10.0.
- pnpm version: 8.12.1.
- MongoDB version: 7.0.2.

### Additional context

This behavior was first discussed with @huyenltnguyen here: https://github.com/freeCodeCamp/freeCodeCamp/issues/52750#issuecomment-1868991073.

I'll try to investigate more in this issue and figure out how to fix it. But, if you have any idea what this might be related to, I would be grateful if you point me in the right direction.

### PR DESCRIPTION
**Pull request by WOLFRIEND**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #52772

### CODE REVIEW

File: curriculum/challenges/english/18-project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions.md [hunk 1]
```diff
@@ -75,29 +75,34 @@ coinPartitions(7);
 
 ```js
 function coinPartitions(divisor) {
-  const partitions = [1];
+  const partitions = Array(divisor + 1).fill(0);
+  partitions[0] = 1;
+
+  for (let i = 1; i <= divisor; i++) {
```
Comment (jeremylt) [inline curriculum/challenges/english/18-project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions.md:81]: Minor - are there words we can use for what i, j, k mean? People do look at these sometimes so I try to make them clear

**Review by jeremylt — COMMENTED** [review]
Can you note where you grabbed the solution ideas from in the PR? I'm trying to be better about that for the solutions I research

**Comment by WOLFRIEND** [comment]
> Can you note where you grabbed the solution ideas from in the PR? I'm trying to be better about that for the solutions I research

Actually, it's kind of a "mixed bag" of solutions found, mostly in other languages and transpiled in JS, plus AI, plus some own refactoring. So, unfortunately there is no "source of truth".
If you found a more elegant solution, that would be great.

File: curriculum/challenges/english/18-project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions.md (hunk 1 above)
Comment (WOLFRIEND) [reply curriculum/challenges/english/18-project-euler/project-euler-problems-1-to-100/problem-78-coin-partitions.md:81]: Hmm, actually:
`i` - is the same as `n` in the initial solution, so it's like `currentN`.
`j` - could be `pentagonalTerm`.
`k` - could be `pentagonalIndex`.

**Comment by jeremylt** [comment]
Hmm, unfortunately using AI here destroys any attribution to the original sources, which I don't love.

**Comment by jeremylt** [comment]
here's a faster version built from
https://blog.dreamshire.com/project-euler-78-solution/

```js
// compute pentagonal numbers per generating function
const pentagonalNumbers = Array(251)
  .fill(0)
  .flatMap((_, i) => i ? [i * (3 * i - 1) / 2, i * (3 * i - 1) / 2 + i] : []);

function coinPartitions(divisor) {
  // helper data
  const signs = [1, 1, -1, -1];

  // compute partition counts until we find a multiple of divisor
  const partitions = Array(divisor + 1).fill(0);
  partitions[0] = 1;
  for (let i = 1; partitions[i - 1] > 0; i++) {
    // compute next partition count
    for (let j = 0; pentagonalNumbers[j] <= i; j++) {
      partitions[i] += partitions[i - pentagonalNumbers[j]] * signs[j % 4];
    }
    partitions[i] = partitions[i] % divisor;
    if (partitions[i] < 0) partitions[i] += divisor; // positive mod
    // return when found
    if (partitions[i] === 0) return i;
  }
}
```

**Comment by WOLFRIEND** [comment]
> ```js
> pentagonalNumbers
> ```

Yeah, this approach is faster. Should I update my PR or you want to create your own?

**Comment by jeremylt** [comment]
Feel free to use that solution in this PR

**Comment by WOLFRIEND** [comment]
> Feel free to use that solution in this PR

Thanks for the help 👍 .
I've updated the PR.

**Review by jeremylt — APPROVED** [review]
Looks good to me!

**Review by gikf — APPROVED** [review]

**Pull request merged**